content/concepts/consensus-network/invariant-checking.md

```diff
@@ -0,0 +1,404 @@
+# Invariant Checking
+
+This article provides a high-level overview of Invariant Checking, why it exist, how it works, and lists active invariant checks.
+
+When building applications on the XRP Ledger, it is crucial to understand the idea of Invariant Checks so as not to be surprised by specific error codes.
+
+## Introduction
+
+Invariant checking is a protection enhancement to the XRP Ledger and reinforces the critical properties of the XRP Ledger.
+
+Invariant checks should not trigger, but they ensure the XRP Ledger's integrity from bugs yet to be discovered or even created.
+
+
+## Why it Exists
+
+Nevertheless, why do we need invariant checks?
+
+- The overarching code for the XRP Ledger is complicated and vast; therefore, there is a high potential for code to execute incorrectly.
+- The cost of incorrectly executing a transaction is high and not acceptable by any standards.
+
+
+## How it Works
+
+A second layer of code runs automatically and in real-time after each transaction completes. It then examines the changes it made for correctness before the results are committed to the ledger. After every transaction runs, the invariant checker runs. Problematic transactions are marked with a **tecINVARIANT_FAILED** result code and are included in the ledger as having done nothing.
+
+The invariant checker is then invariantly checked to ensure the invariant checker's failure is not due to a bug. If this second level check fails, transactions are marked with a **tefINVARIANT_FAILED** result code. The transaction is in its final state and is not included in the ledger.
```

Comment:
Most people, even software engineers, won't easily understand what the phrase "invariantly checked" means here. I checked with Nik and I think I can suggest a better explanation of what's going on here:

```suggestion
To include the transaction in the ledger with a `tec`-class code, some minimal processing is necessary. If this minimal processing still breaks an invariant, the transaction fails with the code `tefINVARIANT_FAILED` instead, and is not included in the ledger at all.
```

Also, despite the "tef" code, tefINVARIANT_FAILED is not a guaranteed final state. Since it only happens if there's a bug (likely a pretty specific one) there's a chance that the changes to the ledger could change the circumstances and cause the bug not to arise again, so resubmitting the transaction could conceivably result in success or a "normal" tec failure. So I left out the "in its final state" sentence.